src/internal.c

```diff
@@ -6224,6 +6232,10 @@ int SetSSL_CTX(WOLFSSL* ssl, WOLFSSL_CTX* ctx, int writeDup)
     ssl->ctx     = ctx; /* only for passing to calls, options could change */
     /* Don't change version on a SSL object that has already started a
      * handshake */
+#if defined(WOLFSSL_HAPROXY)
+    ssl->initial_ctx     = ctx; /* Save access to session key materials */
+    wolfSSL_CTX_up_ref(ctx);
```

Comment:
Please check the return for errors.

Reply:
And only *DO* the assignment when the reference counter is successfully incremented.